stockfish.py

```diff
@@ -0,0 +1,144 @@
+from typing import Dict, List, Optional, Union
+
+
+class Stockfish:
+    """ Integrates the Stockfish chess engine with Python. """
+
+    def __init__(
+        self, path: str = None, depth: int = 2, param: Dict[str, Union[str, int]] = None
+    ):
+        if param is None:
+            param = {}
+        if path is None:
+            path = "stockfish"
+        self.stockfish = subprocess.Popen(
+            path, universal_newlines=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE
+        )
+        self.depth = str(depth)
+
+        # Tell the engine to use the UCI format
+        self.__put("uci")
+
+        default_param = {
+            "Write Debug Log": "false",
+            "Contempt": 0,
+            "Min Split Depth": 0,
+            "Threads": 1,
+            "Ponder": "false",
+            "Hash": 16,
+            "MultiPV": 1,
+            "Skill Level": 20,
+            "Move Overhead": 30,
+            "Minimum Thinking Time": 20,
+            "Slow Mover": 80,
+            "UCI_Chess960": "false",
+        }
+
+        default_param.update(param)
+        self.param = default_param
+        for name, value in list(default_param.items()):
+            self.__set_option(name, value)
+
+        self.__start_new_game()
+
+    def change_thinking_time(milliseconds: int) -> None:
+        """ Changes the thinking time of the chess engine in milliseconds """
+        self.__set_option("Minimum Thinking Time", milliseconds)
+
+    def change_difficulty(difficulty: int) -> None:
+        """ Changes the difficulty of the engine, STILL UNSURE WHAT THIS MEANS """
+        self.__set_option("Skill Level", difficulty)
+
+    def __start_new_game(self) -> None:
+        self.__put("ucinewgame")
+        self.__isready()
+
+    def __put(self, command: str) -> None:
+        """ Internal function for writing commands to the engine """
+        self.stockfish.stdin.write(command + "\n")
+        self.stockfish.stdin.flush()
+
+    def __set_option(self, optionname: str, value: Union[str, int]) -> None:
+        """ Will set an option in the chess engine """
+        self.__put("setoption name %s value %s" % (optionname, str(value)))
+        stdout = self.__isready()
+        if stdout.find("No such") >= 0:
+            print("Unable to set option %s" % optionname)
+
+    def __isready(self) -> str:
+        """ Tests to see if the engine is ready to go """
+        self.__put("isready")
+        while True:
+            text = self.stockfish.stdout.readline().strip()
+            if text == "readyok":
+                return text
+
+    def __go(self):
+        """ will search through the number of plies designated by self.depth"""
+        self.__put("go depth %s" % self.depth)
+
+    @staticmethod
+    def __convert_move_list_to_str(moves: List[str]) -> str:
+        result = ""
+        for move in moves:
+            result += move + " "
+        return result.strip()
+
+    def set_position(self, moves: List[str] = None) -> None:
```

Comment:
Set position is in algebraic notation? Do we switch up between algebraic and fen?

Reply:
That's one thing that's tough. We have to set stockfish's board using fen notation, but we have to manually change the FEN notation whenever we process a move. Once we have the new-FEN, we pass that to stockfish and have it search for more moves. And stockfish does not return a new FEN notation for a move it thinks we should do, but rather it returns the algebraic notation, i.e. `e2e4`